dev/TabView/TabView.cpp

```diff
@@ -0,0 +1,260 @@
+{
+    winrt::IDependencyProperty property = args.Property();
+
+    if (property == s_TabWidthModeProperty)
+    {
+        UpdateTabWidths();
+    }
+}
+
+winrt::AutomationPeer TabView::OnCreateAutomationPeer()
+{
+    return winrt::make<TabViewAutomationPeer>(*this);
+}
+
+void TabView::OnLoaded(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    UpdateTabContent();
+}
+
+void TabView::OnScrollViewerLoaded(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        m_scrollDecreaseButton.set(SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollDecreaseButton").as<winrt::RepeatButton>());
+        if (auto decreaseButton = m_scrollDecreaseButton.get())
+        {
+            m_scrollDecreaseClickRevoker = decreaseButton.Click(winrt::auto_revoke, { this, &TabView::OnScrollDecreaseClick });
+        }
+
+        m_scrollIncreaseButton.set(SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollIncreaseButton").as<winrt::RepeatButton>());
+        if (auto increaseButton = m_scrollIncreaseButton.get())
+        {
+            m_scrollIncreaseClickRevoker = increaseButton.Click(winrt::auto_revoke, { this, &TabView::OnScrollIncreaseClick });
+        }
+
+        if (SharedHelpers::IsRS2OrHigher())
+        {
+            if (auto scrollContentPresenter = SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollContentPresenter").as<winrt::ScrollContentPresenter>())
+            {
+                scrollContentPresenter.TabFocusNavigation(winrt::KeyboardNavigationMode::Once);
+            }
+        }
+    }
+
+    UpdateTabWidths();
+}
+
+void TabView::OnSizeChanged(const winrt::IInspectable& sender, const winrt::SizeChangedEventArgs& args)
+{
+    UpdateTabWidths();
+}
+
+void TabView::OnItemsChanged(winrt::IInspectable const& item)
+{
+    if (auto args = item.as< winrt::IVectorChangedEventArgs>())
+    {
+        if (args.CollectionChange() == winrt::CollectionChange::ItemRemoved && Items().Size() > 0)
+        {
+            if (SelectedIndex() == (int32_t)args.Index())
+            {
+                // Find the closest tab to select instead.
+                int startIndex = (int)args.Index();
+                if (startIndex >= (int)Items().Size())
+                {
+                    startIndex = (int)Items().Size() - 1;
+                }
+                int index = startIndex;
+
+                do
+                {
+                    auto nextItem = ContainerFromIndex(index).as<winrt::ListViewItem>();
+
+                    if (nextItem && nextItem.IsEnabled() && nextItem.Visibility() == winrt::Visibility::Visible)
+                    {
+                        // We need to wait until OnSelectionChanged fires to change the selection, otherwise it will get lost.
+                        m_isTabClosing = true;
+                        m_indexToSelect = index;
+                        break;
```

Comment:
Recommend: Pack these two bits of related state together into a std::optional<int> m_indexToSelectOnSelectionChanged;

then in OnSelectionChanged:
```c++
if (m_indexToSelectOnSelectionChanged)
{
    SelectedIndex = m_indexToSelectOnSelectionChanged.value();
    m_indexToSelectOnSelectionChanged = {};
}
```

Reply:
Done.